### ISSUE TITLE
Fortran cross-reference update (ticket-0)

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#427]
**Originally by _RolfRabenseifner_ on 2014-05-13 10:58:17 -0500**

---
## Description

The changes are for consistency reasons.
Level is ticket-0.
## History

These cross references in MPI-3.0 were not correct enough.

Although these changes are part of the Chapters 3 (Pt-to-pt), 4 (Datatypes),
11 (One-sided), 13 (I/O), and 17.1 (Fortran),
the maintenance of all this cross-references should be done by
the language-binding working group and chapter committee.
## Extended Scope

None. (No need to add these changes to the erratas.)
## Proposed Solution

A) **MPI-3.0 Section 3.7.2 page 52 lines 19-25 read**

  Advice to users. 
  To prevent problems with the argument 
  copying and register optimization done by Fortran compilers, 
  please note the hints in Sections 17.1.10-17.1.20,
  ~~especially in Sections 17.1.12 and 17.1.13 on pages 626-629 
  about "Problems Due to Data Copying and Sequence Association 
  with Subscript Triplets" and "Vector Subscripts",
  and in Sections 17.1.16 to 17.1.19 on pages 631 to 642 about 
  "Optimization Problems", "Code Movements and Register Optimization",
  "Temporary Data Movements" and "Permanent Data Movements"~~. 
  (End of advice to users.)

-_but should read_*

   Advice to users. 
   To prevent problems with the argument 
   copying and register optimization done by Fortran compilers, 
   please note the hints in Sections 17.1.10-17.1.20.
   **Especially Sections 
   17.1.12, 17.1.13 and 17.1.16 to 17.1.19 
   describe conflicts between using 
   nonblocking routines
   together with Fortran,
   that may cause undefined behavior;
   in some cases, solutions are provided.**
   (End of advice to users.)

B) **MPI-3.0 Section 3.9 page 78 lines 30-36 read**

  Advice to users. 
  To prevent problems with the argument 
  copying and register optimization done by Fortran compilers, 
  please note the hints in Sections 17.1.10-17.1.20,
  ~~especially in Sections 17.1.12 and 17.1.13 on pages 626-629 
  about "Problems Due to Data Copying and Sequence Association 
  with Subscript Triplets" and "Vector Subscripts",
  and in Sections 17.1.16 to 17.1.19 on pages 631 to 642 about 
  "Optimization Problems", "Code Movements and Register Optimization",
  "Temporary Data Movements" and "Permanent Data Movements"~~. 
  (End of advice to users.)

-_but should read_*

   Advice to users. 
   To prevent problems with the argument 
   copying and register optimization done by Fortran compilers, 
   please note the hints in Sections 17.1.10-17.1.20.
   **Especially Sections 
   17.1.12, 17.1.13 and 17.1.16 to 17.1.19 
   describe conflicts between using 
   persistent requests
   together with Fortran,
   that may cause undefined behavior;
   in some cases, solutions are provided.**
   (End of advice to users.)

C) **MPI-3.0 Section 4.1.5 page 103 lines 40-46 read**

  Advice to users. 
  To prevent problems with the argument 
  copying and register optimization done by Fortran compilers, 
  please note the hints in Sections 17.1.10-17.1.20,
  ~~especially in Sections 17.1.12 and 17.1.13 on pages 626-629 
  about "Problems Due to Data Copying and Sequence Association 
  with Subscript Triplets" and "Vector Subscripts",
  and in Sections 17.1.16 to 17.1.19 on pages 631 to 642 about 
  "Optimization Problems", "Code Movements and Register Optimization",
  "Temporary Data Movements" and "Permanent Data Movements"~~. 
  (End of advice to users.)

-_but should read_*

   Advice to users. 
   To prevent problems with the argument 
   copying and register optimization done by Fortran compilers, 
   please note the hints in Sections 17.1.10-17.1.20.
   **Especially Sections 
   17.1.12 to 17.1.14 and 17.1.16 to 17.1.19
   describe conflicts between using 
   MPI_BOTTOM
   together with Fortran,
   that may cause undefined behavior;
   in some cases, solutions are provided.**
   (End of advice to users.)

D) **MPI-3.0 Section 11.7.4 page 465 lines 3-10 read**

  To prevent problems with the argument 
  copying and register optimization done by Fortran compilers, 
  please note the hints in Sections 17.1.10-17.1.20,
  ~~especially in Sections 17.1.12 and 17.1.13 on pages 626-629 
  about "Problems Due to Data Copying and Sequence Association 
  with Subscript Triplets" and "Vector Subscripts",
  and in Sections 17.1.16 to 17.1.19 on pages 631 to 642 about 
  "Optimization Problems", "Code Movements and Register Optimization",
  "Temporary Data Movements" and "Permanent Data Movements"~~.
  Sections "Solutions" to 
  "VOLATILE" 
  on pages 634-639 discuss several solutions for the problem in this example. 

-_but should read_*

   To prevent problems with the argument 
   copying and register optimization done by Fortran compilers, 
   please note the hints in Sections 17.1.10-17.1.20.
   **Especially Sections 
   17.1.12, 17.1.13 and 17.1.16 to 17.1.19 
   describe conflicts between using 
   RMA operations or direct access to shared memory windows
   together with Fortran,
   that may cause undefined behavior;
   in some cases, solutions are provided.**
   Sections "Solutions" **through** 
   **"The (Poorly Performing) Fortran VOLATILE Attribute"** 
   on pages 634-639 discuss several solutions for the problem in this example.

E) **MPI-3.0 Section 13.4.1 page 507 lines 6-12 read**

  Advice to users. 
  To prevent problems with the argument 
  copying and register optimization done by Fortran compilers, 
  please note the hints in Sections 17.1.10-17.1.20,
  ~~especially in Sections 17.1.12 and 17.1.13 on pages 626-629 
  about "Problems Due to Data Copying and Sequence Association 
  with Subscript Triplets" and "Vector Subscripts",
  and in Sections 17.1.16 to 17.1.19 on pages 631 to 642 about 
  "Optimization Problems", "Code Movements and Register Optimization",
  "Temporary Data Movements" and "Permanent Data Movements"~~. 
  (End of advice to users.)

-_but should read_*

   Advice to users. 
   To prevent problems with the argument 
   copying and register optimization done by Fortran compilers, 
   please note the hints in Sections 17.1.10-17.1.20.
   **Especially Sections 
   17.1.12, 17.1.13 and 17.1.16 to 17.1.19 
   describe conflicts between using 
   nonblocking I/O routines
   together with Fortran,
   that may cause undefined behavior;
   in some cases, solutions are provided.**
   (End of advice to users.)

F) **MPI-3.0 Section 13.4.5, page 524 lines 24-28 read**
- Split collective routines must specify a buffer
   in both the begin and end routines. By specifying
   the buffer that receives data in the end routine,
   we can avoid the problems described in "A Problem
   with Code Movements and Register Optimization,"
   Section 17.1.17 on page 632, but not all of the 
   problems described in 
   Section 17.1.16 on page 631.

-_but should read_*
- Split collective routines must specify a buffer
   in both the begin and end routines. By specifying
   the buffer that receives data in the end routine,
   we can avoid the problems described in "A Problem
   with Code Movements and Register Optimization,"
   Section 17.1.17 on page 632, but not all of the 
   problems described in 
   Section__s 17.1.12, 17.1.13 and__ 17.1.16 on 
   page__s 626, 629, and__ 631.

G) **MPI-3.0 Section 17.1.12, page 627 line 22 reads**

   If MPI_SUBARRAYS_SUPPORTED equals .FALSE.:

-_but should read_*

   If MPI_SUBARRAYS_SUPPORTED equals .FALSE.:

   **In this case, the use of Fortran arrays with subscript triplets
   as actual choice buffer arguments in any nonblocking
   MPI operation (which also includes persistent request,  
   and split collectives), 
   or together with MPI_BOTTOM
   may cause undefined behavior.
   They may, however, be used in blocking MPI operations.**

Remark for the Fortran WG: "may case" is used instead of
"will cause", because in the case of vector subscripts 
that are simply contiguous, there should not be a conflict.

H) **MPI-3.0 Section 17.1.13, page 629 lines 19-21 read**

   Arrays with a vector subscript must not be used 
   as actual choice buffer arguments in any nonblocking 
   ~~or split collective~~ MPI operations. 
   They may, however, be used in blocking MPI operations.

-_but should read_*

   **Fortran**
   arrays with a vector subscript must not be used 
   as actual choice buffer arguments in any nonblocking
   MPI operation, **or together with MPI_BOTTOM**. 
   They may, however, be used in blocking MPI operations.

I) **MPI-3.0 Section 4.1.5 page 102 lines 30-33 read**

   This initial address zero is indicated by the constant
   MPI_BOTTOM. Thus, a datatype can specify the absolute
   address of the entries in the communication buffer,
   in which case the buf argument is passed the value
   MPI_BOTTOM.

-_but should read_*

   This initial address zero is indicated by the constant
   MPI_BOTTOM. Thus, a datatype can specify the absolute
   address of the entries in the communication buffer,
   in which case the buf argument is passed the value
   MPI_BOTTOM.
   **Note that in Fortran MPI_BOTTOM is not usable for 
   initialization or assignment, see Section 2.5.4.**

Remark for the Fortran WG: Section 17.1.14 is okay.

Remark for the Fortran WG:
Section 17.1.15 on "Fortran Derived Types" is not related
to the areas of,  nonblocking routines, persistent requests,
MPI_BOTTOM, RMA operations or direct access to shared 
memory Windows, and nonblocking I/O routines.

Remark for the Fortran WG: Section 17.1.16 is okay:
It mentions already all relevant conflict areas.

J) **MPI-3.0 Section 17.1.17 page 633 line 29 reads**

   One-sided Communication

-_but should read_*

   **Persistent Requests**

   **With persistent requests, the buffer argument is hidden
   from the MPI_START and MPI_STARTALL calls, i.e., the Fortran compiler may
   move buffer accesses across the MPI_START or MPI_STARTALL call,
   similar to the MPI_WAIT call as described in the
   previous section on nonblocking operations.**

   One-sided Communication

Remark for the Fortran WG: Section 17.1.18 may be okay.

K) **MPI-3.0 Section 17.1.19 page 642 lines 14-16 read**
- Nonblocking MPI operations ~~(communication, one-sided, I/O)~~
   if the internally used pointers to the buffers are not 
   updated by the Fortran runtime, or if within an MPI
   process, the data movement is executed in parallel 
   with the MPI operation.

-_but should read_*
- **All** nonblocking MPI operations
   if the internally used pointers to the buffers are not 
   updated by the Fortran runtime, or if within an MPI
   process, the data movement is executed in parallel 
   with the MPI operation.
## Alternative Solutions

None.
## Impact on Implementations

None required.
## Impact on Applications / Users

None.
## Entry for the Change Log

None.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2014-05-14 10:56:47 -0500**

---

Updated based on the review of Jeff Squyres.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2014-05-14 10:59:14 -0500**

---

Typo correction

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2014-06-02 12:31:24 -0500**

---

Here's feedback from Cisco's internal review of this ticket:
- A-E: We are against these kinds of unnecessary cross references.  The overall sections that must be read are already cited; narrowing down the range is not useful and has already led to some editorial angst.
- G: Seems like a separate thing and should be its own ticket.
- H: Seems like a separate thing and should be its own ticket.
- I: This is redundant with 2.5.4 and should not be here.
- J: Seems like a separate thing and should be its own ticket.
- K: The intent of this clarification is good, but the sentence is still a grammatical train wreck.  If you're going to clarify this sentence, please also fix the sentence itself.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2014-06-03 14:43:34 -0500**

---

Replying to jsquyres:

> Here's feedback from Cisco's internal review of this ticket:
> - A-E: We are against these kinds of unnecessary cross references.  The overall sections that must be read are already cited; narrowing down the range is not useful and has already led to some editorial angst.

The current text is incorrect. The proposed text is a correction with
the smallest difference to the existing text.

> - G: Seems like a separate thing and should be its own ticket.

Yes, it is independent from A-E and may be voted separately.

> - H: Seems like a separate thing and should be its own ticket.

Yes, it is independent from A-E and may be voted separately.

> - I: This is redundant with 2.5.4 and should not be here.

The wording at 2.5.4 is redundant and one wanted to have this redundancy.
We did for all constants, except MPI_BOTTOM, which is the most important from all
and the one that is used as reference.
- See MPI-3.0 p32:30-32
  Note that in Fortran MPI_STATUS_IGNORE
  and MPI_STATUSES_IGNORE are objects like MPI_BOTTOM (not usable for initialization or
  assignment). See Section 2.5.4.
- See MPI-3.0 p380:31-33
  Advice to implementors. MPI_ERRCODES_IGNORE in Fortran is a special type of
  constant, like MPI_BOTTOM. See the discussion in Section 2.5.4 on page 15. (End of
  advice to implementors.)
- And so on for MPI_IN_PLACE (p145:1-2),
  MPI_UNWEIGHTED and MPI_WEIGHTS_EMPTY (p298:3-4 and p300:6-7)
- Exception: MPI_ARGV_NULL and MPI_ARGVS_NULL (no reference to 2.5.4 in Chapter 10),

Therefore, the introduction of MPI_BOTTOM should have at least the same quality
as all these other special constants

> - J: Seems like a separate thing and should be its own ticket.

Yes, it is independent from A-E and may be voted separately.

> - K: The intent of this clarification is good, but the sentence is still a grammatical train wreck.  If you're going to clarify this sentence, please also fix the sentence itself.

No idea to make it better and not significantly longer, because it is part of a list.
You may do it better.

You may vote A-E and F-K independently.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2014-06-05 08:27:23 -0500**

---

I will not allow voting on portions of tickets.  That path leads to madness; I cannot track portions of tickets properly.

Rolf: this ticket was formally read, but I suspect it will be voted down today because it must be voted on as a whole.

I suspect that portions of this ticket should be broken up and either be real ticket 0 (i.e., no voting required), or advanced as errata in Japan.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2014-06-05 10:33:29 -0500**

---

Ticket failed to get per-vote quorum (needed 14.3 votes): 0 yes, 13 no, 6 abstain

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2014-08-12 10:57:20 -0500**

---

I split off the different errata proposal to allow separate single voting on each topic.

Items F-K went to tickets #440, #441, #442, #443, #444, #445.

Here in #427, the alternative with SHORT text is added. The forom should first decide (and document) between SHORT and LONG before voting the ticket at all.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2014-08-13 02:28:20 -0500**

---

I reinstalled the old description text because all comments are based on this text. Items A-E went now to #446. 
This ticket is therefore obsolete and closed (withdrawn).
